File: Exercise_1.py

# ## Problem1 (https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/)
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No


# // Your code here along with comments explaining your approach
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, inorder, postorder):
        # {inorder values: inorder index}
        self.hmap = {}
        self.idx = len(postorder) - 1
        for i in range(len(inorder)):
            self.hmap[inorder[i]] = i
        self.helper(inorder, postorder, 0, len(inorder) - 1)

    def helper(self, inorder, postorder, start, end):
        # base
        if start > end:
            return None

        # logic
        rootVal = postorder[self.idx]
        root = TreeNode(rootVal)
        self.idx -= 1
        rootIdx = self.hmap.get(rootVal)

        root.right = self.helper(inorder, postorder, rootIdx + 1, end)
        root.left = self.helper(inorder, postorder, start, rootIdx - 1)

        return root


sol = Solution()
inorder = [9, 3, 15, 20, 7]
postorder = [9, 15, 7, 20, 3]

print(sol.buildTree(inorder, postorder))

File: Exercise_1.cpp
// Time Complexity                              :  O(N) where N is the length of the inorder and postorder arrays
// Space Complexity                             :  O(N) due to values stored in the function stack. 
//                                                 O(N) in case of a skewed tree otherwise O(H) where H is 
//                                                  the height of the tree.
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private: 
    int pi;
    
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        int n = inorder.size();
        if(n == 1) return new TreeNode(inorder[0]);
        pi = n - 1;
        return bTree(inorder, postorder,0,n-1);
    }
    
    TreeNode* bTree(vector<int>& inorder, vector<int>& postorder, int start, int end) {
        if(start > end) return NULL;
        
        TreeNode *nn = new TreeNode(postorder[pi--]);
        int pos;
        for(int i=start;i<=end;i++) {
            if(inorder[i] == nn->val) {
                pos = i; 
                break;
            }
        }
        
        nn->right = bTree(inorder,postorder,pos+1,end);
        nn->left = bTree(inorder,postorder,start,pos-1);
        return nn;
    }
};